src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt

```diff
@@ -65,15 +68,54 @@ class PrisonerSearchService(@Autowired private val client: PrisonerSearchApiClie
     )
 
   fun findPotentialMatch(matchPrisonersRequest: MatchPrisonersRequest): List<PotentialMatch> {
-    var list = mutableListOf<PotentialMatch>()
-    try {
-      var count = matchPrisonersRequest.pncNumber?.toInt() ?: 0
-      repeat(count) {
-        list.add(PotentialMatch(firstName = "FirstName", lastName = "LastName", dateOfBirth = LocalDate.now(), null, null))
-      }
-      return list
-    } catch (ex: NumberFormatException) {
-      return list
+
+    var map = hashMapOf<String, PotentialMatch>()
+
+    var listMatchPrisonerResponse =
+      this.getCandidateMatches(matchPrisonersRequest.prisonNumber, matchPrisonersRequest.pncNumber)
+    listMatchPrisonerResponse.forEach {
+      map[it.prisonerNumber ?: "" + "_" + it.lastName + "_" + it.firstName] = PotentialMatch(
+        firstName = it.firstName,
+        lastName = it.lastName,
+        dateOfBirth = it.dateOfBirth,
+        pncNumber = it.pncNumber,
+        prisonNumber = it.prisonerNumber
+      )
+    }
+
+    var searchByPncNumber = SearchByNameAndDateOfBirth(
+      firstName = matchPrisonersRequest.firstName,
+      lastName = matchPrisonersRequest.lastName,
+      dateOfBirth = matchPrisonersRequest.dateOfBirth,
+      pncNumber = matchPrisonersRequest.pncNumber
+    )
+    var listByPncNumber = client.matchPrisonerByNameAndDateOfBirth(searchByPncNumber)
+    listByPncNumber.forEach {
+      map[it.prisonerNumber + "_" + it.lastName + "_" + it.firstName] = PotentialMatch(
+        firstName = it.firstName,
+        lastName = it.lastName,
+        dateOfBirth = it.dateOfBirth,
+        pncNumber = it.pncNumber,
+        prisonNumber = it.prisonerNumber
+      )
+    }
+
+    var searchByNameAndDateOfBirth = SearchByNameAndDateOfBirth(
+      firstName = matchPrisonersRequest.firstName,
+      lastName = matchPrisonersRequest.lastName,
+      dateOfBirth = matchPrisonersRequest.dateOfBirth
+    )
+    var listByNameAndDateOfBirth = client.matchPrisonerByNameAndDateOfBirth(searchByNameAndDateOfBirth)
+
+    listByNameAndDateOfBirth.forEach {
+      map[it.prisonerNumber + "_" + it.lastName + "_" + it.firstName] = PotentialMatch(
+        firstName = it.firstName,
+        lastName = it.lastName,
+        dateOfBirth = it.dateOfBirth,
+        pncNumber = it.pncNumber,
+        prisonNumber = it.prisonerNumber
+      )
     }
+    return map.values.toList()
```

Comment:
Can't you add all these results to a list? and then use `distinctBy` to remove duplicates based on prison number?

Reply:
What about scenario when I have multiple record back without prison number

Reply:
sorted